### ISSUE TITLE
Python kernels crash when started from Docker with CMD without a shell

### ISSUE THREAD

**Issue by thewtex** [jupyter/notebook#334]
When the jupyter notebook is started as in this Dockerfile: https://gist.github.com/thewtex/baba0c786f8dcdc9c743
the Python kernel repeatedly crashes.

However, when started with a shell as here:
https://gist.github.com/thewtex/1316b3a2b4f3c170686b
the kernels do not crash.

If this is known and expected, it would be helpful to have a note in the "Known Issues" section of the documents, and have an example of what works in jupyter/notebook/Dockerfile.

**Comment by Carreau** [comment]
Do you have any tracebacks or error messages ?

**Comment by thewtex** [comment]
The kernel crashes when I start a notebook.  There in not anything informative in jupyter's stdout, stderr.  Is there a log file I should be looking at?

**Comment by Carreau** [comment]
you can try to start with --debug to get more info on stdout. But I have little knowledge on how docker works. @rgbkrk @jhamrick ?

**Comment by jhamrick** [comment]
Hmm, yeah, I don't know why it would be crashing -- I've never had an issue with that. A few thoughts, though not sure if any of these would actually be related:

1. You probably don't want to run the container as root (i.e. you should probably define a user in it and then set the working directory to that user's home directory)
2. You might want to have it listen to `0.0.0.0` rather than `*`

I would have a look at how the `jupyter/singleuser` container is set up, which is used with JupyterHub (so you won't need e.g. the singleuser script, but the stuff it does in terms of creating a user and stuff might be relevant):

https://github.com/jupyter/dockerspawner/blob/master/singleuser/Dockerfile

I suppose though that docker container also uses `sh` to ultimately start the server so maybe that's why it works. Now that I think about it a bit more, I'm not actually sure I've tried to run the notebook directly from a container, it's always been from some script that did some additional setup first.

**Comment by rgbkrk** [comment]
That's definitely a known issue (https://github.com/ipython/ipython/issues/7062). We're not sure if the issue is in pexpect or elsewhere, but we only ever have this issue in Docker. PID reaping?

/cc @parente

**Comment by jakirkham** [comment]
The fact that calling iPython within bash or a simple bash script solves the problem certainly suggest that PID reaping is the issue.

For those unfamiliar, this is a pretty nice description of the problem ( https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem ).

Maybe using their image ( https://hub.docker.com/r/phusion/baseimage ) or building some limited parts of it ( https://github.com/phusion/baseimage-docker ) would make for a more robust base image.

**Comment by parente** [comment]
I suspect the same. It's one of the reasons the github.com/jupyter/docker-stacks images all use supervisord as PID 1.

Is the docker image in this repo meant to be used for working in a jupyter notebook server? Or is it meant to be a development image? I thought the latter, in which case this problem isn't really an issue if it doesn't affect development of the notebook project. But if that's true, something should probably get put in the README stating so and directing people to the stacks images instead.

**Comment by jakirkham** [comment]
"...[Supervisod] is not meant to be run as a substitute for init as “process id 1”" ( http://supervisord.org )

**Comment by parente** [comment]
> "...[Supervisod] is not meant to be run as a substitute for init as “process id 1”" ( http://supervisord.org )

Indeed it isn't meant for this use, but in practice with jupyter notebook, it does handle the cases where a kernel spawns a child process (e.g., a Python kernel spawning a py4jvm JVM instance for Spark).

```
2015-07-31 19:22:11,927 CRIT reaped unknown pid 77)
2015-08-07 17:48:45,894 INFO stopped: notebook (exit status 0)
2015-08-07 17:48:45,894 CRIT reaped unknown pid 1078)
2015-08-07 17:48:45,894 CRIT reaped unknown pid 1136)
2015-08-07 17:48:45,903 INFO spawned: 'notebook' with pid 1191
2015-08-07 17:48:46,913 INFO success: notebook entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2015-08-19 13:11:08,553 CRIT reaped unknown pid 1265)
2015-08-19 17:30:40,718 CRIT reaped unknown pid 1297)
```

That said, I'm all for switching to another more official solution over in docker-stacks and will open an issue over there. Need to consider the fact that those images are based on debian:jessie vs ubuntu and what a switch might mean for size. 

I can't speak for what's appropriate over here in jupyter/notebook, though.

**Comment by jakirkham** [comment]
Don't get me wrong. Something doing the reaping is better than nothing doing the reaping. Also, the current solution is better than best theoretical solution as it does already exist. That being said, having a proper init process would be nice.

**Comment by parente** [comment]
Sure. I documented some alternatives over in the docker-stacks issue I opened.

**Comment by thewtex** [comment]
I have tried out tini in my base image:

  https://github.com/thewtex/jupyter-notebook-debian

and it works well.  Adding this to the documentation in this pull request:

  https://github.com/jupyter/notebook/pull/485

should prevent fewer people from tripping over this.

**Comment by parente** [comment]
jupyter/minimal-notebook now uses tini as well (ref https://github.com/jupyter/docker-stacks/pull/38 )

**Comment by minrk** [comment]
closed by #530
